actionview/lib/action_view/template.rb

```diff
@@ -237,39 +237,39 @@ def encode!
       end
     end
 
-    protected
+    # Templates can be used concurrently in threaded environments
+    # so compilation and any instance variable modification must
+    # be synchronized
+    def thread_safe_compile!(view)
+      return if @compiled
 
-      # Compile a template. This method ensures a template is compiled
-      # just once and removes the source after it is compiled.
-      def compile!(view) #:nodoc:
+      @compile_mutex.synchronize do
+        # Any thread holding this lock will be compiling the template needed
+        # by the threads waiting. So re-check the @compiled flag to avoid
+        # re-compilation
         return if @compiled
+        compile!(view)
+      end
+    end
 
-        # Templates can be used concurrently in threaded environments
-        # so compilation and any instance variable modification must
-        # be synchronized
-        @compile_mutex.synchronize do
-          # Any thread holding this lock will be compiling the template needed
-          # by the threads waiting. So re-check the @compiled flag to avoid
-          # re-compilation
-          return if @compiled
-
-          if view.is_a?(ActionView::CompiledTemplates)
-            mod = ActionView::CompiledTemplates
-          else
-            mod = view.singleton_class
-          end
+    # Compile a template. This method ensures a template is compiled
+    # just once and removes the source after it is compiled.
+    def compile!(view = nil) #:nodoc:
```

Comment:
The view is only used to check if we can store the compiled template (which is an instance method) in ActionView::CompiledTemplates. In our code, we always want to store it there. This change will guarantee that when a view is passed to the method, it works like it used to work, but now we can call `compile!` here 
https://github.com/nygrenh/rails/blob/69807105154c8452986b01e2cd4c0364bbc074bb/actionview/lib/action_view/template_eager_loader.rb#L16
without a parameter.

We think that creating an unnecessary view object in the eager loading loop would slow it down and result in a lot of extra garbage collecting.